src/wayland/virtual_keyboard/virtual_keyboard_handle.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    ffi::CString,
+    fmt::Debug,
+    sync::{Arc, Mutex},
+};
+
+use wayland_protocols_misc::zwp_virtual_keyboard_v1::server::zwp_virtual_keyboard_v1::{
+    self, ZwpVirtualKeyboardV1,
+};
+use wayland_server::{
+    backend::{ClientId, ObjectId},
+    protocol::wl_keyboard::{KeyState, KeymapFormat},
+    Client, DataInit, Dispatch, DisplayHandle, Resource,
+};
+use xkbcommon::xkb;
+
+use crate::{
+    input::{keyboard::KeymapFile, Seat, SeatHandler},
+    utils::SERIAL_COUNTER,
+    wayland::seat::{keyboard::for_each_focused_kbds, WaylandFocus},
+};
+
+use super::VirtualKeyboardManagerState;
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Hash)]
+struct SerializedMods {
+    depressed: u32,
+    latched: u32,
+    locked: u32,
+    group: u32,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct VirtualKeyboard {
+    pub instances: Vec<ZwpVirtualKeyboardV1>,
+    modifiers: Option<SerializedMods>,
+    old_keymap: Option<KeymapFile>,
+}
+
+/// Handle to a virtual keyboard instance
+#[derive(Debug, Clone, Default)]
+pub(crate) struct VirtualKeyboardHandle {
+    pub(crate) inner: Arc<Mutex<VirtualKeyboard>>,
+}
+
+impl VirtualKeyboardHandle {
+    pub(super) fn add_instance<D>(&self, instance: &ZwpVirtualKeyboardV1) {
+        let mut inner = self.inner.lock().unwrap();
+        inner.instances.push(instance.clone());
+    }
+}
+
+/// User data of ZwpVirtualKeyboardV1 object
+#[derive(Debug)]
+pub struct VirtualKeyboardUserData<D: SeatHandler> {
+    pub(super) handle: VirtualKeyboardHandle,
+    pub(crate) seat: Seat<D>,
+}
+
+impl<D> Dispatch<ZwpVirtualKeyboardV1, VirtualKeyboardUserData<D>, D> for VirtualKeyboardManagerState
+where
+    D: Dispatch<ZwpVirtualKeyboardV1, VirtualKeyboardUserData<D>>,
+    D: SeatHandler + 'static,
+    <D as SeatHandler>::KeyboardFocus: WaylandFocus,
+{
+    fn request(
+        _data: &mut D,
+        _client: &Client,
+        _: &ZwpVirtualKeyboardV1,
+        request: zwp_virtual_keyboard_v1::Request,
+        data: &VirtualKeyboardUserData<D>,
+        _dh: &DisplayHandle,
+        _data_init: &mut DataInit<'_, D>,
+    ) {
+        match request {
+            zwp_virtual_keyboard_v1::Request::Keymap { format, fd, size } => {
+                if format == 1 {
```

Comment:
I guess the protocol does not mark the format u32 as an enum...

Reply:
Unfortunately not

Reply:
We should probably open an issue for that then where ever this protocol lives

Reply:
@i509VCB This is part of wayland-protocols

Reply:
No? This was posted on the mailing list but never integrated into wayland-protocols. I think purism is the source of the protocol.

Reply:
I guess we should at least note why this is weird.

Something like:
```
This should be wl_keyboard::KeymapFormat::XkbV1, but the protocol does not state the parameter is an enum.
```

Reply:
In addition to mailing list it also has MR open now: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/11

Reply:
The MR indeed has the enum label

Reply:
The alternative would be to use the new version, basically being a forerunner?

Reply:
added comment :)